src/util/arith_tools.h

```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
 #define CPROVER_UTIL_ARITH_TOOLS_H
 
 #include "mp_arith.h"
+#include "optional.h"
+#include "invariant.h"
 
 class exprt;
 class constant_exprt;
 class typet;
 
 // this one will go away
 // returns 'true' on error
+/// \deprecated: use the constant_exprt version instead
 bool to_integer(const exprt &expr, mp_integer &int_value);
 
 // returns 'true' on error
+/// \deprecated: use numeric_cast<mp_integer> instead
 bool to_integer(const constant_exprt &expr, mp_integer &int_value);
 
 // returns 'true' on error
 bool to_unsigned_integer(const constant_exprt &expr, unsigned &uint_value);
 
+/// Numerical cast provides a unified way of converting from one numerical type
+/// to another.
+/// Generic case doesn't exist, this has to be specialized for different types.
+template <typename Target, typename = void>
+struct numeric_castt final
+{
+};
+
+/// Convert expression to mp_integer
+template <>
+struct numeric_castt<mp_integer> final
+{
+  optionalt<mp_integer> operator()(const exprt &expr) const
+  {
+    mp_integer out;
+    if(to_integer(expr, out))
+      return {};
+    return out;
+  }
+};
+
+/// Convert mp_integer or expr to any signed type
+template <typename T>
+struct numeric_castt<T,
+  typename std::enable_if<std::is_integral<T>::value &&
+                          std::is_signed<T>::value>::type>
+{
+  optionalt<T> operator()(const mp_integer &mpi) const
+  {
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
+    static_assert(
+      std::numeric_limits<T>::max() <=
+      std::numeric_limits<decltype(mpi.to_long())>::max() &&
+      std::numeric_limits<T>::min() >=
+      std::numeric_limits<decltype(mpi.to_long())>::min(),
+      "Numeric cast only works for types smaller than long long");
+#else
+    PRECONDITION(
+      std::numeric_limits<T>::max() <=
+        std::numeric_limits<decltype(mpi.to_long())>::max() &&
+      std::numeric_limits<T>::min() >=
+        std::numeric_limits<decltype(mpi.to_long())>::min());
+#endif
+    if(
+      mpi <= std::numeric_limits<T>::max() &&
+      mpi >= std::numeric_limits<T>::min())
+      // to_long converts to long long which is the largest signed numeric type
+      return static_cast<T>(mpi.to_long());
+    else
+      return {};
+  }
+
+  optionalt<T> operator()(const exprt &expr) const
+  {
+    auto mpi_opt = numeric_castt<mp_integer>{}(expr);
+    if(mpi_opt)
+      return numeric_castt<T>{}(*mpi_opt);
+    else
+      return {};
+  }
+};
+
+/// Convert mp_integer to any unsigned type
+/// \tparam T: type to convert to
+/// \param mpi: mp_integer to convert
+/// \return optional integer of type T if conversion is possible,
+///         empty optional otherwise.
+template <typename T>
+struct numeric_castt<T,
+  typename std::enable_if<std::is_integral<T>::value &&
```

Comment:
This seems to be the same function as above, except for the `mpi` function called. Could we refactor?

Reply:
I refactored a bit to have only one structure for the conversion to integral type, but I still have one function for mpi to signed and one function from mpi to unsigned because all the checks are different.

Reply:
Are you sure? I'm pretty sure the checks can be made the same (I show how this is done in my proposal PR)

Reply:
Yes it can be done, I will try that.